src/auth/credential.ts

```diff
@@ -364,116 +263,162 @@ export class RefreshTokenCredential implements Credential {
   }
 }
 
+class RefreshToken {
 
-/**
- * Implementation of Credential that gets access tokens from the metadata service available
- * in the Google Cloud Platform. This authenticates the process as the default service account
- * of an App Engine instance or Google Compute Engine machine.
- */
-export class MetadataServiceCredential implements Credential {
-
-  private readonly httpClient = new HttpClient();
-  private readonly httpAgent?: Agent;
-
-  constructor(httpAgent?: Agent) {
-    this.httpAgent = httpAgent;
-  }
+  public readonly clientId: string;
+  public readonly clientSecret: string;
+  public readonly refreshToken: string;
+  public readonly type: string;
 
-  public getAccessToken(): Promise<GoogleOAuthAccessToken> {
-    const request: HttpRequestConfig = {
-      method: 'GET',
-      url: `http://${GOOGLE_METADATA_SERVICE_HOST}${GOOGLE_METADATA_SERVICE_PATH}`,
-      headers: {
-        'Metadata-Flavor': 'Google',
-      },
-      httpAgent: this.httpAgent,
-    };
-    return requestAccessToken(this.httpClient, request);
+  /*
+   * Tries to load a RefreshToken from a path. Throws if the path doesn't exist or the
+   * data at the path is invalid.
+   */
+  public static fromPath(filePath: string): RefreshToken {
+    try {
+      return new RefreshToken(JSON.parse(fs.readFileSync(filePath, 'utf8')));
+    } catch (error) {
+      // Throw a nicely formed error message if the file contents cannot be parsed
+      throw new FirebaseAppError(
+        AppErrorCodes.INVALID_CREDENTIAL,
+        'Failed to parse refresh token file: ' + error,
+      );
+    }
   }
-}
-
 
-/**
- * ApplicationDefaultCredential implements the process for loading credentials as
- * described in https://developers.google.com/identity/protocols/application-default-credentials
- */
-export class ApplicationDefaultCredential implements FirebaseCredential {
-  private credential_: Credential;
+  constructor(json: object) {
+    copyAttr(this, json, 'clientId', 'client_id');
+    copyAttr(this, json, 'clientSecret', 'client_secret');
+    copyAttr(this, json, 'refreshToken', 'refresh_token');
+    copyAttr(this, json, 'type', 'type');
```

Comment:
I am wondering why the `type` property is needed.

Reply:
Yes, it's kind of redundant. It's always going to have the constant value "authorized_user" in this case. I think this type is meant to represent the refresh token JSON file, which usually contains a "type" field. I'd keep it as is, since removal of the property is not really related to this refactor. We can remove it in a future PR if needed.